pandas/io/parsers.py

```diff
@@ -1894,23 +1899,52 @@ def __init__(self, src, **kwds):
 
         ParserBase.__init__(self, kwds)
 
-        encoding = kwds.get("encoding")
-
-        # parsers.TextReader doesn't support compression dicts
-        if isinstance(kwds.get("compression"), dict):
-            kwds["compression"] = kwds["compression"]["method"]
-
-        if kwds.get("compression") is None and encoding:
-            if isinstance(src, str):
-                src = open(src, "rb")
-                self.handles.append(src)
-
-            # Handle the file object with universal line mode enabled.
-            # We will handle the newline character ourselves later on.
-            if hasattr(src, "read") and not hasattr(src, "encoding"):
-                src = TextIOWrapper(src, encoding=encoding, newline="")
-
-            kwds["encoding"] = "utf-8"
+        src = stringify_path(src)
+        if (
+            kwds.get("memory_map", False)
+            and isinstance(src, str)
+            and get_compression_method(kwds.get("compression", None))[0] is None
+        ):
+            # memory-mapped files are directly handled by the TextReader.
+            # compression and file-object were never supported by it.
+            self.source = src
+            self.handles = []
+        else:
+            if kwds.get("memory_map", False):
+                if not isinstance(src, str):
+                    warnings.warn(
+                        "read_csv does support only string-like objects with "
+                        + "memory_map=True and engine='c'. Falling back to "
+                        + "memory_map=False. Please use engine='python' instead.",
+                        RuntimeWarning,
+                    )
+                else:
+                    warnings.warn(
+                        "read_csv does not support compression with "
+                        + "memory_map=True. Falling back to memory_map=False. "
+                        + "Please use memory_map=False instead.",
+                        RuntimeWarning,
+                    )
+            self.source, self.handles = get_handle(
+                src,
+                mode="rb",
+                encoding=kwds.get("encoding", "utf-8"),
+                compression=kwds.get("compression", None),
+                is_text=True,
+                errors=None,
+            )
+            kwds.pop("encoding", None)
+        kwds.pop("memory_map", None)
+        kwds.pop("compression", None)
+
+        # use id() instead of objects as S3FileSystem's __eq__ expects all file objects
+        # to have the attribute 'mode' (TextIOWrapper doesn't has 'mode')
```

Comment:
can we use a sub-class of TextIOWrapper to give the same api? will that make it simpler

Reply:
Do you mean to inherit from `TextIOWrapper` and add a fake `mode` attribute?

I will first try a simpler solution (not sure whether it will work): x == y and y == x  will call different `__eq__` methods, or? Imho `self.source == src ` should work, while `src == self.source` might fail (src could be a S3FileSystem, get_handle will wrap it into an TextIOWrapper: self.source shouldn't be a S3FileSystem)

Reply:
unfortunately that doesn't work. In this case it is imho okay to call `id`to compares the objects as we want to catch cases were `get_handle` did not apply any wrappings.